fabric-entity-tick-events-v1/src/main/java/net/fabricmc/fabric/impl/event/entity/TickEventInternals.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.impl.event.entity;
+
+import java.util.IdentityHashMap;
+import java.util.Map;
+
+import com.google.common.base.Preconditions;
+
+import net.minecraft.entity.Entity;
+import net.minecraft.util.profiler.Profiler;
+
+import net.fabricmc.fabric.api.event.Event;
+import net.fabricmc.fabric.api.event.EventFactory;
+import net.fabricmc.fabric.api.event.v1.EntityTickCallback;
+
+public class TickEventInternals {
+	/** Cache of entity class to {@code Event}. */
+	private static final Map<Class<?>, CascadingEvent<?>> ENTITY_TICK_EVENTS = new IdentityHashMap<>();
```

Comment:
what is the point of identity hash map? a regular one should suffice, or maybe we need concurrent/synchronized hash maps

Reply:
You shouldn't be ticking an entity on the wrong thread in the first place.

I have honestly been wondering if we should have different listeners for ticking on a logical client and logical server

Reply:
hmm, what if there's concurrent registration on client and server threads

Reply:
I think you would usually register this during `onInitialize`. However you could argue the same with concurrent registration for literally every other event in fabric's api